### ISSUE TITLE
Properly support UPDATE semantics in addition to PATCH semantics for 'id' property

### ISSUE THREAD

**Issue by dhermes** [GoogleCloudPlatform/endpoints-proto-datastore#65]
Properly support `UPDATE` semantics in addition to `PATCH` semantics for `'id'` property. (And the `'entityKey'` analogues, throughout).

Currently, since `IdSet` calls `UpdateFromKey`, if an entity does exist, all the properties that weren't set on the entity from the request are "patched" from the version that is stored in the datastore.

If a user wanted to "properly" implement `UPDATE` it would look something like this:

```
  # Using request_message, **not** request_fields
  @MyModel.method(request_message=MyModel.ProtoModel(),
                  path='path/{id}', name='mymodel.update')
  def my_model(self, my_model_proto_request):  # Not an entity
    def update_existing():
      existing = MyModel.get_by_id(my_model_proto_request.id)
      if existing is None:
        raise endpoints.NotFoundException('Not found error')

      # Potentially some other validation logic on
      # my_model_proto_request

      # Remove the id so UpdateFromKey in IdSet doesn't get called when
      # MyModel.FromMessage is called.
      my_model_proto_request.id = None
      my_model = MyModel.FromMessage(my_model_proto_request)
      # Set key explicitly instead of 'id' since that would call IdSet
      my_model.key = existing.key
      my_model.put()
      return my_model

    return ndb.transaction(update_existing)
```

**ADDENDUM**: To really support CRUD, it may either make sense to add `update_method`, `create_method`, etc. class methods on `EndpointsModel` or to provide a way to create hooks into `EndpointsModel.method` with a well-defined and well-documented way to emulate UPDATE instead of PATCH.

/cc @danholevoet @fredsa

**Comment by dhermes** [comment]
@kdeus Any updates here? I know @craigcitro has `protorpc` on GitHub now and ready for PRs.

**Comment by kdeus** [comment]
Yeah, I talked with Craig and the rest of the ProtoRPC team.  The cleaner
fix for this, which has some other benefits, is to make the entire request
available to the server.  Right now, it's only accessible via the Messages
that ProtoRPC has already processed.  Making the raw request available will
give the server a lot more flexibility.  For this case, it'll allow you to
distinguish whether a field was provided or not.

That's the plan right now.  I expect this'll make it into the next App
Engine release, but <blah blah can't comment on future releases> no
guarantees.  :)

**Comment by dhermes** [comment]
Great! Thanks. I wonder if there are any public changes to https://github.com/google/protorpc you could point to?

**Comment by kdeus** [comment]
The changes will be in Google Cloud Endpoints, not in ProtoRPC.

**Comment by dhermes** [comment]
Gotcher. Thanks @kdeus

**Comment by tmst** [comment]
Hey, guys, I'm trying to track down why the fields of an entity's subobject (StructuredProperty) are getting wiped when any of the subobject's fields are being set by the client. I believe it started happening recently and isn't related to anything I've done in my app. Could it be related to this issue?

**Comment by dhermes** [comment]
@tmst It's definitely possible. Can you provide an example / what you expect / how it fails?

**Comment by tmst** [comment]
@dhermes Sure. I just noticed that my first example was a poor one. I'll try to put together a better one in the next day or so.
